### PR TITLE
inplace codebook optimization

### PR DESCRIPTION
**Pull request by theAdamColton**
This is a little bit unwieldy to use, it requires using a lambda like this:
```python
import torch
from torch.optim import SGD
from vector_quantize_pytorch import VectorQuantize

vq = VectorQuantize(dim=512,
                    codebook_size=32,
                    codebook_dim=512,
                    heads=1,
                    ema_update=False,
                    learnable_codebook=True,
                    in_place_codebook_optimizer=lambda *args, **kwargs: SGD(*args, **kwargs, lr=50.0),
                )

x = torch.rand(7, 1, 512, requires_grad=True)

for _ in range(10):
    *_, loss = vq(x)

    print(loss) # This will decrease over time as the vq._codebook.embed is updated
```

Another possibility to get the learnable codebook to work would be for the VectorQuantizer to return a codebook commit loss, which would be some sort of distance like this `((x.detach() - quantize)**2).mean()`. This would not be as efficient because you'd have to pass it through the project_in and project_out an extra time for every in-place update.

### CODE REVIEW

**Comment by lucidrains** [comment]
what would happen if the straight through looked something like

```python
quantize = quantize + (x - x.detach())
```

**Comment by lucidrains** [comment]
i'll think more about this once i get back from europe

**Comment by theAdamColton** [comment]
you'd probably have to also change the commit loss because right now it's being calculated with the detached_quantize

**Comment by lucidrains** [comment]
> you'd probably have to also change the commit loss because right now it's being calculated with the detached_quantize

oh yes, you are correct

should be addressed in 1.6.14!

File: vector_quantize_pytorch/vector_quantize_pytorch.py [hunk 1]
```diff
@@ -797,6 +804,17 @@ def forward(
 
         quantize, embed_ind, distances = self._codebook(x, sample_codebook_temp = sample_codebook_temp)
 
+
+        if should_inplace_optimize and self.training:
+            # One step in-place update
+            ((quantize - x.detach())**2).mean().backward()
```
Comment (lucidrains) [inline vector_quantize_pytorch/vector_quantize_pytorch.py:810]: you could use `F.mse_loss` here?

**Comment by lucidrains** [comment]
@theAdamColton hey, this looks good! left a few comments but otherwise lgtm

have you tested it?

**Comment by lucidrains** [comment]
@theAdamColton accepting, as i think it should work

thank you for this!
